packages/@blockprotocol/graph/src/type-resolver/entity-type-to-typescript.ts

```diff
@@ -0,0 +1,216 @@
+  let compiledSchema = rootSchema ? generateImportStatements() : "";
+
+  // Generate the type for FooProperties, representing the entity's 'own properties' (not links or linked entities)
+  compiledSchema += await compileSchema(
+    { ...schema, title: propertyTypeName },
+    {
+      additionalProperties: false, // @todo add additionalProperties: false to entity type JSON schemas
+      bannerComment: rootSchema ? bannerComment(schema.$id, depth) : "",
+    },
+  );
+
+  // This adds 'type Foo = Entity<FooProperties>;'
+  const entityTypeDefinition = generateEntityDefinition(
+    typeName,
+    propertyTypeName,
+  );
+  compiledSchema += entityTypeDefinition;
+
+  // using a type instead of an interface allows using FooProperties in Entity<FooProperties>
+  // matching group needed in case the compiler has made e.g. FooProperties1 & FooProperties2 to handle inheritance
+  // @todo we need to catch types generated internally by the library via inheritance (allOf), which we don't know about
+  //   either these need to be converted to types too, or we need a different approach to Entity<Properties>
+  compiledSchema = compiledSchema.replace(
+    new RegExp(`export interface (${typeName}Properties\\S*)`, "g"),
+    `export type $1 =`,
+  );
+
+  // temporary hack for the @todo mentioned above
+  // this covers the one type we know is currently generated via inheritance
+  compiledSchema = compiledSchema.replace(
+    /export interface Link/g,
+    "export type Link =",
+  );
+
+  const compiledType = { typeName, typeScriptString: compiledSchema };
+
+  // add the compiled type to our map of already-resolved types in case it's encountered again when following links
+  // eslint-disable-next-line no-param-reassign -- this is intentionally mutated when this function is called
+  resolvedUrisToType[schema.$id] = compiledType;
+
+  if (depth === 0) {
+    return compiledType;
+  }
+
+  // if we're following links, we want to generate various types that can be plugged into subgraph functions
+
+  // keep track of the possible destination types from this entity type for each link type it includes
+  const typeLinkMap: Record<string, string> = {};
+
+  for (const [linkEntityTypeUri, destinationSchema] of Object.entries(
+    schema.links ?? {},
+  )) {
+    const retrieveOrCompileSchemaFromUri = async (
+      uri: string,
+    ): Promise<CompiledType> => {
+      const cachedType = resolvedUrisToType[uri];
+      if (cachedType) {
+        return {
+          typeName: cachedType.typeName,
+          typeScriptString: "",
+        };
+      }
+      const typeSchema = await fetchAndValidateEntityType(uri as VersionedUri);
+      return _jsonSchemaToTypescript(typeSchema, depth - 1, resolvedUrisToType);
+    };
+
+    // get the schema for the linkEntity and possible rightEntities for this link, from this entity
+    const [linkEntityType, ...rightEntityTypes] = await Promise.all([
+      retrieveOrCompileSchemaFromUri(linkEntityTypeUri),
```

Comment:
If you use the `typedEntries` helper instead of `Object.entries` I think it should fix the problems I'm guessing you're encountering here with `VersionedUri` and such becoming untyped (or poorly typed)

Reply:
Have fixed all this